state/execution.go

```diff
@@ -417,15 +416,44 @@ func (blockExec *BlockExecutor) Commit(
 	)
 
 	// Update mempool.
+	ch <- mempoolUpdate{block: block, state: state.Copy(), abciResponse: abciResponse, err: err}
+
+	return res.RetainHeight, nil
+}
+
+// updates the mempool with the latest state asynchronously.
+// it begins flushing the mempool immediately, and then updates the mempool
+// after the new block state is ready.
+func (blockExec *BlockExecutor) asyncUpdateMempool(ch chan mempoolUpdate) {
+	blockExec.mempool.Lock()
+	defer blockExec.mempool.Unlock()
+
+	// while mempool is Locked, flush to ensure all async requests have completed
+	// in the ABCI app before Commit.
+	err := blockExec.mempool.FlushAppConn()
+	if err != nil {
+		blockExec.logger.Error("client error during mempool.FlushAppConn, flushing mempool", "err", err)
+		blockExec.mempool.Flush()
+		return
+	}
+
+	updateData := <-ch
+	if updateData.err != nil {
+		return
+	}
+
 	err = blockExec.mempool.Update(
-		block.Height,
-		block.Txs,
-		abciResponse.TxResults,
-		TxPreCheck(state),
-		TxPostCheck(state),
+		updateData.block.Height,
+		updateData.block.Txs,
+		updateData.abciResponse.TxResults,
+		TxPreCheck(updateData.state),
+		TxPostCheck(updateData.state),
 	)
-
-	return res.RetainHeight, err
+	if err != nil {
+		blockExec.logger.Error("client error during mempool.Update, flushing mempool", "err", err)
+		blockExec.mempool.Flush()
```

Comment:
This is new, right? Why it's important to flush the mempool?

Reply:
Before we would halt, sorry forgot to flag in a comment. I feel like we only have two options,  dump the mempool or replicate existing functionality and make an error channel to halt consensus

Reply:
Any caller of function `Commit` in line 395 above will panic if `Commit` doesn't return a `nil` error. This is for safety. I'd feel more comfortable if we didn't modify that behavior here, so `err != nil` in line 445, I'd panic here with the error text of line 446.

Reply:
Sounds good!